File: php/src/Gts.php
<?php
// SPDX-FileCopyrightText: 2026 Blackcat Informatics Inc. <paudley@blackcatinformatics.ca>
// SPDX-License-Identifier: MIT OR Apache-2.0

declare(strict_types=1);

namespace Gmeow\Gts;

use FFI;
use FFI\CData;
use InvalidArgumentException;
use RuntimeException;

final class Gts
{
    private const CDEF_TEMPLATE = <<<'CDEF'
typedef unsigned char uint8_t;
typedef unsigned int uint32_t;
@GTS_SIZE_T_TYPEDEF@

typedef enum gts_status {
  GTS_STATUS_OK = 0,
  GTS_STATUS_INVALID_ARGUMENT = 1,
  GTS_STATUS_IO = 2,
  GTS_STATUS_PARSE = 3,
  GTS_STATUS_DIAGNOSTIC = 4,
  GTS_STATUS_INTERNAL = 5,
  GTS_STATUS_PANIC = 6
} gts_status;

typedef struct gts_buffer {
  uint8_t *data;
  size_t len;
  size_t capacity;
} gts_buffer;

typedef struct gts_error gts_error;

uint32_t gts_abi_version(void);
const char *gts_version(void);

void gts_buffer_free(gts_buffer *buffer);
void gts_error_free(gts_error *error);
const char *gts_error_code(const gts_error *error);
const char *gts_error_message(const gts_error *error);

gts_status gts_build_metadata_json(gts_buffer *out, gts_error **error);
gts_status gts_capabilities_json(gts_buffer *out, gts_error **error);
gts_status gts_read_json(const uint8_t *data, size_t len, gts_buffer *out, gts_error **error);
gts_status gts_verify_json(const uint8_t *data, size_t len, gts_buffer *out, gts_error **error);
gts_status gts_to_nquads(const uint8_t *data, size_t len, gts_buffer *out, gts_error **error);
gts_status gts_from_nquads(const char *text, size_t len, gts_buffer *out, gts_error **error);

gts_status gts_files_pack(const char *const *paths, size_t path_count, gts_buffer *out, gts_error **error);
gts_status gts_files_unpack(
  const uint8_t *data,
  size_t len,
  const char *dest,
  uint32_t flags,
  gts_buffer *out,
  gts_error **error
);
gts_status gts_files_diff_json(
  const uint8_t *data,
  size_t len,
  const char *directory,
  gts_buffer *out,
  gts_error **error
);
CDEF;

    private FFI $ffi;

    public function __construct(?string $library = null)
    {
        $this->ffi = FFI::cdef(self::cdef(), $library ?? self::defaultLibrary());
    }

    public static function load(?string $library = null): self
    {
        return new self($library);
    }

    public static function defaultLibrary(): string
    {
        $fromEnv = getenv('GTS_LIBGTS');
        if (is_string($fromEnv) && $fromEnv !== '') {
            return $fromEnv;
        }
        return match (PHP_OS_FAMILY) {
            'Windows' => 'gts.dll',
            'Darwin' => 'libgts.dylib',
            default => 'libgts.so',
        };
    }

    public function abiVersion(): int
    {
        return (int) $this->ffi->gts_abi_version();
    }

    public function version(): string
    {
        return self::copyCString($this->ffi->gts_version());
    }

    public function buildMetadataJson(): string
    {
        return $this->callBuffer(
            'gts_build_metadata_json',
            fn (CData $out, CData $error): int => (int) $this->ffi->gts_build_metadata_json($out, $error)
        );
    }

    public function capabilitiesJson(): string
    {
        return $this->callBuffer(
            'gts_capabilities_json',
            fn (CData $out, CData $error): int => (int) $this->ffi->gts_capabilities_json($out, $error)
        );
    }

    public function readJson(string $data): string
    {
        return $this->withBytes(
            $data,
            'uint8_t',
            fn (CData $dataPtr, int $len): string => $this->callBuffer(
                'gts_read_json',
                fn (CData $out, CData $error): int => (int) $this->ffi->gts_read_json(
                    $dataPtr,
                    $len,
                    $out,
                    $error
                )
            )
        );
    }

    public function verifyJson(string $data): string
    {
        return $this->withBytes(
            $data,
            'uint8_t',
            fn (CData $dataPtr, int $len): string => $this->callBuffer(
                'gts_verify_json',
                fn (CData $out, CData $error): int => (int) $this->ffi->gts_verify_json(
                    $dataPtr,
                    $len,
                    $out,
                    $error
                )
            )
        );
    }

    public function toNQuads(string $data): string
    {
        return $this->withBytes(
            $data,
            'uint8_t',
            fn (CData $dataPtr, int $len): string => $this->callBuffer(
                'gts_to_nquads',
                fn (CData $out, CData $error): int => (int) $this->ffi->gts_to_nquads(
                    $dataPtr,
                    $len,
                    $out,
                    $error
                )
            )
        );
    }

    public function fromNQuads(string $text): string
    {
        return $this->withBytes(
            $text,
            'char',
            fn (CData $textPtr, int $len): string => $this->callBuffer(
                'gts_from_nquads',
                fn (CData $out, CData $error): int => (int) $this->ffi->gts_from_nquads(
                    $textPtr,
                    $len,
                    $out,
                    $error
                )
            )
        );
    }

    /**
     * @param list<string> $paths
     */
    public function filesPack(array $paths): string
    {
        $paths = array_values($paths);
        [$pathPointers, $pathBuffers] = $this->nativePathList($paths);
        $pathCount = count($paths);

        return $this->callBuffer(
            'gts_files_pack',
            function (CData $out, CData $error) use ($pathPointers, $pathBuffers, $pathCount): int {
                return (int) $this->ffi->gts_files_pack($pathPointers, $pathCount, $out, $error);
            }
        );
    }

    public function filesUnpack(string $data, string $destination, int $flags = GtsUnpackFlags::NONE): string
    {
        $destinationPtr = $this->nativeString($destination, 'destination');

        return $this->withBytes(
            $data,
            'uint8_t',
            fn (CData $dataPtr, int $len): string => $this->callBuffer(
                'gts_files_unpack',
                fn (CData $out, CData $error): int => (int) $this->ffi->gts_files_unpack(
                    $dataPtr,
                    $len,
                    $destinationPtr,
                    $flags,
                    $out,
                    $error
                )
            )
        );
    }

    public function filesDiffJson(string $data, string $directory): string
    {
        $directoryPtr = $this->nativeString($directory, 'directory');

        return $this->withBytes(
            $data,
            'uint8_t',
            fn (CData $dataPtr, int $len): string => $this->callBuffer(
                'gts_files_diff_json',
                fn (CData $out, CData $error): int => (int) $this->ffi->gts_files_diff_json(
                    $dataPtr,
                    $len,
                    $directoryPtr,
                    $out,
                    $error
                )
            )
        );
    }

    /**
     * @param callable(CData, CData): int $call
     */
    private function callBuffer(string $operation, callable $call): string
    {
        $out = $this->ffi->new('gts_buffer');
        $error = $this->ffi->new('gts_error *');

        try {
            $status = $call(FFI::addr($out), FFI::addr($error));
            if ($status !== GtsStatus::OK) {
                throw $this->buildException($operation, $status, $error);
            }
            if (!FFI::isNull($error)) {
                throw $this->buildException($operation, GtsStatus::INTERNAL, $error);
            }
            return self::copyBuffer($out);
        } finally {
            $this->ffi->gts_buffer_free(FFI::addr($out));
        }
    }

    private static function cdef(): string
    {
        $sizeT = PHP_OS_FAMILY === 'Windows'
            ? 'typedef unsigned long long size_t;'
            : 'typedef unsigned long size_t;';
        return str_replace('@GTS_SIZE_T_TYPEDEF@', $sizeT, self::CDEF_TEMPLATE);
    }

    /**
     * @param callable(CData, int): string $call
     */
    private function withBytes(string $data, string $elementType, callable $call): string
    {
        $len = strlen($data);
        $buffer = $this->ffi->new(sprintf('%s[%d]', $elementType, max(1, $len)));
        if ($len > 0) {
            FFI::memcpy($buffer, $data, $len);
        }
        return $call($buffer, $len);
    }

    /**
     * @param list<string> $paths
     * @return array{0: CData, 1: list<CData>}
     */
    private function nativePathList(array $paths): array
    {
        if ($paths === []) {
            throw new InvalidArgumentException('Path list must not be empty.');
        }

        $pathBuffers = [];
        $pathPointers = $this->ffi->new(sprintf('const char *[%d]', count($paths)));
        foreach ($paths as $index => $path) {
            if (!is_string($path)) {
                throw new InvalidArgumentException(sprintf('Path at index %d is not a string.', $index));
            }
            $pathBuffers[$index] = $this->nativeString($path, sprintf('source[%d]', $index));
            $pathPointers[$index] = $this->ffi->cast('char *', $pathBuffers[$index]);
        }

        return [$pathPointers, $pathBuffers];
    }

    private function nativeString(string $value, string $name): CData
    {
        if (str_contains($value, "\0")) {
            throw new InvalidArgumentException(sprintf('%s contains a NUL byte.', $name));
        }
        $len = strlen($value);
        $buffer = $this->ffi->new(sprintf('char[%d]', $len + 1));
        if ($len > 0) {
            FFI::memcpy($buffer, $value, $len);
        }
        $buffer[$len] = "\0";
        return $buffer;
    }

    private function buildException(string $operation, int $status, CData $error): GtsException
    {
        $code = '';
        $detail = '';

        try {
            if (!FFI::isNull($error)) {
                $code = self::copyCString($this->ffi->gts_error_code($error));
                $detail = self::copyCString($this->ffi->gts_error_message($error));
            }
        } finally {
            if (!FFI::isNull($error)) {
                $this->ffi->gts_error_free($error);
            }
        }

        return new GtsException($operation, $status, $code, $detail);
    }

    private static function copyBuffer(CData $buffer): string
    {
        $len = self::checkedLength((int) $buffer->len);
        if ($len === 0) {
            return '';
        }
        if (FFI::isNull($buffer->data)) {
            throw new RuntimeException('C ABI returned a null data pointer with non-zero length.');
        }
        return FFI::string($buffer->data, $len);
    }

    private static function copyCString(mixed $value): string
    {
        if (is_string($value)) {
            return $value;
        }
        if ($value === null || FFI::isNull($value)) {
            return '';
        }
        return FFI::string($value);
    }

    private static function checkedLength(int $len): int
    {
        if ($len < 0) {
            throw new RuntimeException(sprintf('C ABI returned a negative buffer length: %d.', $len));
        }
        return $len;
    }
}

File: dotnet/Gmeow.Gts/Gts.cs
// SPDX-FileCopyrightText: 2026 Blackcat Informatics Inc. <paudley@blackcatinformatics.ca>
// SPDX-License-Identifier: MIT OR Apache-2.0

using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;
using System.Text;

namespace Gmeow.Gts;

public enum GtsStatus
{
    Ok = 0,
    InvalidArgument = 1,
    Io = 2,
    Parse = 3,
    Diagnostic = 4,
    Internal = 5,
    Panic = 6,
}

[Flags]
public enum GtsUnpackFlags : uint
{
    None = 0,
    IncludeSuppressed = 1u << 0,
    AllowSymlinks = 1u << 1,
    AllowSpecial = 1u << 2,
    SameOwner = 1u << 3,
    PreserveSetid = 1u << 4,
}

public sealed class GtsException : Exception
{
    public GtsException(string operation, GtsStatus status, string code, string detail)
        : base(Format(operation, status, code, detail))
    {
        Operation = operation;
        Status = status;
        Code = code;
        Detail = detail;
    }

    public string Operation { get; }
    public GtsStatus Status { get; }
    public string Code { get; }
    public string Detail { get; }

    private static string Format(string operation, GtsStatus status, string code, string detail)
    {
        var message = $"{operation} failed with {status}";
        if (!string.IsNullOrEmpty(code))
        {
            message += $" ({code})";
        }
        if (!string.IsNullOrEmpty(detail))
        {
            message += $": {detail}";
        }
        return message;
    }
}

public static class Gts
{
    public static uint AbiVersion => NativeMethods.gts_abi_version();

    public static string Version => CopyUtf8(NativeMethods.gts_version());

    public static string BuildMetadataJson()
    {
        return CallString("gts_build_metadata_json", NativeMethods.gts_build_metadata_json);
    }

    public static string CapabilitiesJson()
    {
        return CallString("gts_capabilities_json", NativeMethods.gts_capabilities_json);
    }

    public static string ReadJson(byte[] data)
    {
        ArgumentNullException.ThrowIfNull(data);
        return CallString("gts_read_json", (out NativeMethods.GtsBuffer buffer, out IntPtr error) =>
            NativeMethods.gts_read_json(data, ToUIntPtr(data.Length), out buffer, out error));
    }

    public static string VerifyJson(byte[] data)
    {
        ArgumentNullException.ThrowIfNull(data);
        return CallString("gts_verify_json", (out NativeMethods.GtsBuffer buffer, out IntPtr error) =>
            NativeMethods.gts_verify_json(data, ToUIntPtr(data.Length), out buffer, out error));
    }

    public static string ToNQuads(byte[] data)
    {
        ArgumentNullException.ThrowIfNull(data);
        return CallString("gts_to_nquads", (out NativeMethods.GtsBuffer buffer, out IntPtr error) =>
            NativeMethods.gts_to_nquads(data, ToUIntPtr(data.Length), out buffer, out error));
    }

    public static byte[] FromNQuads(string text)
    {
        ArgumentNullException.ThrowIfNull(text);
        byte[] textBytes = Encoding.UTF8.GetBytes(text);
        return CallBytes("gts_from_nquads", (out NativeMethods.GtsBuffer buffer, out IntPtr error) =>
            NativeMethods.gts_from_nquads(textBytes, ToUIntPtr(textBytes.Length), out buffer, out error));
    }

    public static byte[] FilesPack(IEnumerable<string> paths)
    {
        ArgumentNullException.ThrowIfNull(paths);
        var nativePaths = new List<IntPtr>();
        try
        {
            foreach (string path in paths)
            {
                ArgumentNullException.ThrowIfNull(path);
                nativePaths.Add(Marshal.StringToCoTaskMemUTF8(path));
            }
            IntPtr[] rawPaths = nativePaths.ToArray();
            return CallBytes("gts_files_pack", (out NativeMethods.GtsBuffer buffer, out IntPtr error) =>
                NativeMethods.gts_files_pack(rawPaths, ToUIntPtr(rawPaths.Length), out buffer, out error));
        }
        finally
        {
            foreach (IntPtr path in nativePaths)
            {
                Marshal.FreeCoTaskMem(path);
            }
        }
    }

    public static string FilesUnpack(byte[] data, string destination, GtsUnpackFlags flags = GtsUnpackFlags.None)
    {
        ArgumentNullException.ThrowIfNull(data);
        ArgumentNullException.ThrowIfNull(destination);
        IntPtr nativeDestination = Marshal.StringToCoTaskMemUTF8(destination);
        try
        {
            return CallString("gts_files_unpack", (out NativeMethods.GtsBuffer buffer, out IntPtr error) =>
                NativeMethods.gts_files_unpack(
                    data,
                    ToUIntPtr(data.Length),
                    nativeDestination,
                    (uint)flags,
                    out buffer,
                    out error));
        }
        finally
        {
            Marshal.FreeCoTaskMem(nativeDestination);
        }
    }

    public static string FilesDiffJson(byte[] data, string directory)
    {
        ArgumentNullException.ThrowIfNull(data);
        ArgumentNullException.ThrowIfNull(directory);
        IntPtr nativeDirectory = Marshal.StringToCoTaskMemUTF8(directory);
        try
        {
            return CallString("gts_files_diff_json", (out NativeMethods.GtsBuffer buffer, out IntPtr error) =>
                NativeMethods.gts_files_diff_json(data, ToUIntPtr(data.Length), nativeDirectory, out buffer, out error));
        }
        finally
        {
            Marshal.FreeCoTaskMem(nativeDirectory);
        }
    }

    private delegate GtsStatus BufferCall(out NativeMethods.GtsBuffer buffer, out IntPtr error);

    private static string CallString(string operation, BufferCall call)
    {
        return CallBuffer(operation, call, CopyUtf8);
    }

    private static byte[] CallBytes(string operation, BufferCall call)
    {
        return CallBuffer(operation, call, CopyBytes);
    }

    private static T CallBuffer<T>(string operation, BufferCall call, Func<NativeMethods.GtsBuffer, T> copy)
    {
        NativeMethods.GtsBuffer buffer = default;
        IntPtr error = IntPtr.Zero;
        GtsStatus status = call(out buffer, out error);
        if (status != GtsStatus.Ok)
        {
            throw BuildException(operation, status, error);
        }
        if (error != IntPtr.Zero)
        {
            throw BuildException(operation, GtsStatus.Internal, error);
        }
        try
        {
            return copy(buffer);
        }
        finally
        {
            NativeMethods.gts_buffer_free(ref buffer);
        }
    }

    private static GtsException BuildException(string operation, GtsStatus status, IntPtr error)
    {
        string code = string.Empty;
        string detail = string.Empty;
        try
        {
            if (error != IntPtr.Zero)
            {
                code = CopyUtf8(NativeMethods.gts_error_code(error));
                detail = CopyUtf8(NativeMethods.gts_error_message(error));
            }
        }
        finally
        {
            if (error != IntPtr.Zero)
            {
                NativeMethods.gts_error_free(error);
            }
        }
        return new GtsException(operation, status, code, detail);
    }

    private static string CopyUtf8(NativeMethods.GtsBuffer buffer)
    {
        byte[] bytes = CopyBytes(buffer);
        return Encoding.UTF8.GetString(bytes);
    }

    private static byte[] CopyBytes(NativeMethods.GtsBuffer buffer)
    {
        int length = CheckedLength(buffer.Len);
        if (length == 0)
        {
            return Array.Empty<byte>();
        }
        var bytes = new byte[length];
        Marshal.Copy(buffer.Data, bytes, 0, length);
        return bytes;
    }

    private static string CopyUtf8(IntPtr value)
    {
        return value == IntPtr.Zero ? string.Empty : Marshal.PtrToStringUTF8(value) ?? string.Empty;
    }

    private static UIntPtr ToUIntPtr(int value)
    {
        if (value < 0)
        {
            throw new ArgumentOutOfRangeException(nameof(value), value, "Length must be non-negative.");
        }
        return new UIntPtr((uint)value);
    }

    private static int CheckedLength(UIntPtr value)
    {
        ulong length = value.ToUInt64();
        if (length > int.MaxValue)
        {
            throw new InvalidOperationException($"C ABI output is too large for a managed byte array: {length} bytes.");
        }
        return (int)length;
    }
}
